### ISSUE TITLE
Allow conditions that check for the cabal(-install?) version used.

### ISSUE THREAD

**Issue by AndreasPK** [haskell/cabal#10386]
**Describe the feature request**

The gist of it is that I have a .cmm file in a project that should be linked into the final executable.

Pre-3.12 we used a workaround to deal with https://github.com/haskell/cabal/issues/8639 where we would add the cmm file to `ghc-options` and that worked as it would add Foo.cmm to the link step invocation of ghc which would then compile and linke the resulting .o file into the executable as well.

With >= Cabal 3.12 this results in linker errors as cabal hands the cmm derived .o file to the linker now as it always should have *and* the workaround adds it another time, resulting in duplicate symbols.

Ideally I would work around this using something like `if cabal < 3.12: ghc-options: Foo.cmm` but there does not seem to be a obvious way to do so.

Concretely I have:

```
executable run-it
    main-is:          Run.hs
    cmm-sources:      Trampoline.cmm

    -- Workaround Cabal < 3.12 since Cabal doesn't link cmm-sources into executables
    ghc-options:      Trampoline.cmm
    build-depends:    base,
                      unix,
                      mmap
    ghc-options:      -Wall -Wno-unticked-promoted-constructors
    default-language: Haskell2010
```

In this specific case the package is typically only built via a script so I can add a flag that corresponds to a cabal version predicate that I compute in the bash script. But having the ability to do this in the cabal file itself seems a lot cleaner and not all that far fetched.

**Comment by philderbeast** [comment]
I asked for a similar thing for projects with #8933.

**Comment by dcoutts** [comment]
The theory is that .cabal files declare the version of the Cabal spec that they follow, and then tools interpret that into the behaviours. So .cabal files don't talk about the version of a particular build tool in use (stack, cabal-install, Setup.hs whatever).

So arguably that bux fix ought to be conditional on the declared version of cabal in the package's .cabal file. That is the bug fix should be a change in specified behaviour between Cabal spec version X and X+1, and then .cabal files can either follow the old or new behaviour.

Allowing "if spec( < 3.12)" is arguably not completely implausible, but now it changes the interpretation of .cabal files from "I follow this version of the spec" into "I follow some base version of the spec but tell me a version of the spec and I'll try to adapt". In that case, how does a build tool know what range it can support, and what behaviours it should apply. Currently build tools can say "if the package is pre spec-vX then I'll use this behaviour, else that", but if the package supports a range, then what behaviour should the tool use?

**Comment by philderbeast** [comment]
Between `cabal-3.8` and `cabal-3.10` project imports changed from being "project root" relative to being "importing file" relative. With Updo, I worked around that with a [`CABAL_RELATIVITY`](https://github.com/cabalism/updo#4-package-groups) flag as project files, unlike package description files, are not versioned nor do they allow `if impl(cabal ...)` conditionals.

**Comment by AndreasPK** [comment]
One suggestion that I was made ware of would be to:

* Consider the old behaviour, and perhaps cmm-sources as a whole, a bug and therefore deprecate `cmm-sources` for Cabal specs before 3.12.
* Add a warning to cabal-check to check for the use of cmm-sources with older specs.

This would ensure well defined behavior going forward, prevent packages using cmm-sources making their way onto hackage and would be reasonably easy to implement. The downside of this approach is that it removes the ability to use a spec <3.12 while using cmm-sources. Which might be alright. I was told there are essentially no packages on hackage currently using this combination anyway.

**Comment by Mikolaj** [comment]
@ffaf1 notices this would make exact-printing and parsing (even) harder. There was a reason the .cabal file starts with the version (in particular, to make parsing error messages better). But maybe it's worth it.

**Comment by tfausak** [comment]
Can you explain how this would make parsing or rendering harder? I don't see how that's the case. 

I agree with Andreas here:

> the predicate would only check the version of the _interpreter_ of the spec

So you could have `cabal-version: 3.0` and then in that package description you could say `if cabal(>= 3.2) …`. You wouldn't be setting the _spec_ version to 3.2 within that conditional. Instead you'd be saying that the conditional is only relevant for Cabal(-install) 3.2 and later. 

We already have `impl` for conditionals on the compiler, like `if impl(ghc >= 9.2) …`. Perhaps this should be something like `if build(cabal >= 3.2) …`? That way you could also add `if build(mcabal) …` for targeting MicroHS's version of Cabal.

**Comment by Bodigrim** [comment]
> But maybe it's worth it.

I don't think so, an additional axis of conditionals is not a can of worms worth opening. As an escape hatch people can use `build-type: Custom`. For the original problem above one can set `cabal-version: 3.12` so that older Cabals do not pick up the package. 

Version of Cabal-the-executable is not a great predictor of which bugs/features are there and which are not (things get backported and/or get patched by third parties); they should not be encouraged to be used for such purposes. An additional axis of conditionals would entail something like `tested-with-build-tool: Cabal == 3.12, Cabal == 3.14` and a cultural shift in the community to test routinely against multiple versions of Cabal.

(As a Hackage trustee, I really dread such prospect)

**Comment by sol** [comment]
The given use case, I would describe as "working around a Cabal bug".

Personally, I think, if a user wants to work around this Cabal bug, I hope this would work:

```cabal
cabal-version: 3.12
```

(or if that doesn't work for some reason,  `setup-depends: Cabal >= 3.12`, as @Bodigrim hinted)

As for this feature possibly being useful in other situations, personally I would want to see a convincing use case (not a bug workaround, not a theoretical use case, but a real use case that a real user has right now) to even remotely consider this.  But even then, again personally for me, it's a far stretch.

Addressing the underlaying cabal bug that motivated this issue of course makes sense. (even for older `cabal-version` specs, provided it's possible to do so in a non-breaking way).

**Comment by sol** [comment]
> That way you could also add `if build(mcabal) …` for targeting MicroHS's version of Cabal.

If this would indeed help MicroHs support, then I would be all for it.

But:
- Wouldn't you generally prefer `if impl(mhs >= ...)` over `if build(mcabal) …`?
- If we want more (friendly) competition in the ecosystem, I think keeping the package specification simple in itself is a feature.  I imagine that this will benefit `mcabal`, and everybody else who wants to experiment with new approaches. (and frankly speaking, I think it will also benefit cabal itself)
